File: node/src/tunnel.ts
/**
 * Tunnel local agents to the internet so the cloud engine can reach them.
 *
 * Uses cloudflared (trycloudflare.com, no signup) as primary,
 * falls back to ngrok, bore, or localtunnel.
 */

import { spawn, ChildProcess } from 'child_process';
import { createInterface, ReadLineOptions } from 'readline';
import { platform } from 'os';
import { EOL } from 'os';

/** Check if a URL points to localhost or a private network. */
export function isLocalUrl(url: string): boolean {
  return (
    url.includes('localhost') ||
    url.includes('127.0.0.1') ||
    url.includes('0.0.0.0') ||
    url.includes('192.168.') ||
    url.includes('10.')
  );
}

/** Extract port number from a URL like http://localhost:8047/path. */
export function extractPort(url: string): number | null {
  const m = url.match(/:(\d+)/);
  return m ? parseInt(m[1], 10) : null;
}

export class TunnelError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'TunnelError';
  }
}

// ── Regex patterns ─────────────────────────────────────────────────

const CLOUDFLARED_RE = /https:\/\/[a-zA-Z0-9.-]+\.trycloudflare\.com/;
const NGROK_RE = /url=https:\/\/[a-zA-Z0-9.-]+\.ngrok(?:-free)?\.(?:app|io)/;
const BORE_RE = /listening at (bore\.pub:\d+)/;
const LOCALTUNNEL_RE = /your url is: (https:\/\/[a-zA-Z0-9.-]+\.loca\.lt)/;

// ── Provider definitions ───────────────────────────────────────────

interface Provider {
  binary: string;
  command(bin: string, port: number): string[];
  extractUrl(line: string): string | null;
}

const PROVIDERS: Provider[] = [
  {
    binary: 'cloudflared',
    command: (bin, port) => [bin, 'tunnel', '--url', `http://localhost:${port}`],
    extractUrl: (line) => CLOUDFLARED_RE.exec(line)?.[0] ?? null,
  },
  {
    binary: 'ngrok',
    command: (bin, port) => [bin, 'http', String(port), '--log=stdout'],
    extractUrl: (line) => {
      const m = NGROK_RE.exec(line);
      return m ? m[0].replace('url=', '') : null;
    },
  },
  {
    binary: 'bore',
    command: (bin, port) => [bin, 'local', String(port), '--to', 'bore.pub'],
    extractUrl: (line) => {
      const m = BORE_RE.exec(line);
      return m ? `http://${m[1]}` : null;
    },
  },
  {
    binary: 'lt',
    command: (bin, port) => [bin, '--port', String(port)],
    extractUrl: (line) => {
      const m = LOCALTUNNEL_RE.exec(line);
      return m ? m[1] : null;
    },
  },
];

// ── Which helper (no dependency) ────────────────────────────────────

function whichSync(name: string): string | null {
  // Simple PATH lookup
  const pathEnv = process.env.PATH || '';
  const dirs = pathEnv.split(platform() === 'win32' ? ';' : ':');
  const extensions = platform() === 'win32' ? ['.exe', '.cmd', '.bat', ''] : [''];
  for (const dir of dirs) {
    for (const ext of extensions) {
      const fullPath = `${dir}/${name}${ext}`;
      try {
        require('fs').accessSync(fullPath, require('fs').constants.X_OK);
        return fullPath;
      } catch {
        // Not here
      }
    }
  }
  return null;
}

// ── Tunnel class ───────────────────────────────────────────────────

export class Tunnel {
  private process: ChildProcess | null = null;
  publicUrl: string | null = null;

  /**
   * Start a tunnel and return the public URL.
   * Throws TunnelError if no provider is available or timeout.
   */
  start(localPort: number): Promise<string> {
    return this._tryProviders(localPort);
  }

  /** Stop the tunnel process. */
  stop(): void {
    if (this.process && this.process.exitCode === null) {
      try {
        if (platform() === 'win32') {
          this.process.kill();
        } else {
          // Kill the process group on Unix
          const pid = this.process.pid;
          if (pid) {
            process.kill(-pid, 'SIGTERM');
            // Give it 5s then SIGKILL
            setTimeout(() => {
              try { process.kill(-pid, 'SIGKILL'); } catch { /* gone */ }
            }, 5000).unref();
          }
        }
      } catch {
        // Process already gone
      }
    }
    this.process = null;
  }

  /** Async dispose (for `await using`). */
  async [Symbol.asyncDispose](): Promise<void> {
    this.stop();
  }

  private async _tryProviders(localPort: number): Promise<string> {
    const errors: string[] = [];

    for (const provider of PROVIDERS) {
      const binPath = whichSync(provider.binary);
      if (!binPath) continue;

      try {
        const url = await this._tryProvider(binPath, localPort, provider);
        if (url) return url;
      } catch (e) {
        errors.push(`${provider.binary}: ${(e as Error).message}`);
        this.stop();
      }
    }

    throw new TunnelError(
      'No tunnel tool found. Install one:\n' +
        '  macOS:  brew install cloudflared\n' +
        '  Linux:  curl -L https://github.com/cloudflare/cloudflared/releases/latest/download/cloudflared-linux-amd64 -o /usr/local/bin/cloudflared && chmod +x /usr/local/bin/cloudflared\n' +
        '  Or:     brew install ngrok && ngrok config add-authtoken <token>',
    );
  }

  private _tryProvider(
    binPath: string,
    localPort: number,
    provider: Provider,
  ): Promise<string | null> {
    return new Promise((resolve, reject) => {
      const args = provider.command(binPath, localPort);
      const cmd = args[0];
      const cmdArgs = args.slice(1);

      const child = spawn(cmd, cmdArgs, {
        stdio: ['ignore', 'pipe', 'ignore'],
        detached: platform() !== 'win32',
      });

      this.process = child;

      let settled = false;
      const timeout = setTimeout(() => {
        if (!settled) {
          settled = true;
          this.stop();
          reject(new TunnelError('Tunnel did not provide a public URL within 15 seconds'));
        }
      }, 15000);

      const rl = createInterface({ input: child.stdout! });

      rl.on('line', (line: string) => {
        if (settled) return;
        const url = provider.extractUrl(line);
        if (url) {
          settled = true;
          clearTimeout(timeout);
          rl.close();
          this.publicUrl = url;
          resolve(url);
        }
      });

      child.on('close', (code) => {
        if (!settled) {
          settled = true;
          clearTimeout(timeout);
          this.stop();
          reject(new TunnelError(`Tunnel process exited with code ${code}`));
        }
      });

      child.on('error', (err) => {
        if (!settled) {
          settled = true;
          clearTimeout(timeout);
          this.stop();
          reject(new TunnelError(`Failed to start tunnel: ${err.message}`));
        }
      });
    });
  }
}

File: rust/src/tunnel.rs
//! Tunnel local agents to the internet so the cloud engine can reach them.
//!
//! Uses cloudflared (trycloudflare.com, no signup) as primary,
//! falls back to ngrok, bore, or localtunnel.

use regex::Regex;
use std::fmt;
use std::io::{BufRead, BufReader};
use std::process::{Child, Command, Stdio};
use std::sync::LazyLock;
use std::time::{Duration, Instant};

/// Check if a URL points to localhost or a private network.
pub fn is_local_url(url: &str) -> bool {
    url.contains("localhost")
        || url.contains("127.0.0.1")
        || url.contains("0.0.0.0")
        || url.contains("192.168.")
        || url.contains("10.")
}

/// Extract port number from a URL like http://localhost:8047/path.
pub fn extract_port(url: &str) -> Option<u16> {
    let re = Regex::new(r":(\d+)").ok()?;
    re.captures(url)
        .and_then(|caps| caps.get(1))
        .and_then(|m| m.as_str().parse::<u16>().ok())
}

// ── Regex patterns for URL extraction ──────────────────────────────

static CLOUDFLARED_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"https://[a-zA-Z0-9.-]+\.trycloudflare\.com").unwrap());

static NGROK_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"url=https://[a-zA-Z0-9.-]+\.ngrok(?:-free)?\.(?:app|io)").unwrap()
});

static BORE_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"listening at (bore\.pub:\d+)").unwrap());

static LOCALTUNNEL_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"your url is: (https://[a-zA-Z0-9.-]+\.loca\.lt)").unwrap());

// ── Error type ──────────────────────────────────────────────────────

#[derive(Debug)]
pub struct TunnelError {
    pub message: String,
}

impl fmt::Display for TunnelError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.message)
    }
}

impl std::error::Error for TunnelError {}

// ── Provider definitions ───────────────────────────────────────────

struct Provider {
    binary: &'static str,
    command: fn(&str, u16) -> Vec<String>,
    extract_url: fn(&str) -> Option<String>,
}

fn cloudflared_cmd(bin: &str, port: u16) -> Vec<String> {
    vec![
        bin.to_string(),
        "tunnel".to_string(),
        "--url".to_string(),
        format!("http://localhost:{}", port),
    ]
}

fn ngrok_cmd(bin: &str, port: u16) -> Vec<String> {
    vec![
        bin.to_string(),
        "http".to_string(),
        port.to_string(),
        "--log=stdout".to_string(),
    ]
}

fn bore_cmd(bin: &str, port: u16) -> Vec<String> {
    vec![
        bin.to_string(),
        "local".to_string(),
        port.to_string(),
        "--to".to_string(),
        "bore.pub".to_string(),
    ]
}

fn localtunnel_cmd(bin: &str, port: u16) -> Vec<String> {
    vec![bin.to_string(), "--port".to_string(), port.to_string()]
}

fn extract_cloudflared(line: &str) -> Option<String> {
    CLOUDFLARED_RE.find(line).map(|m| m.as_str().to_string())
}

fn extract_ngrok(line: &str) -> Option<String> {
    NGROK_RE.find(line).map(|m| m.as_str().replace("url=", ""))
}

fn extract_bore(line: &str) -> Option<String> {
    BORE_RE
        .captures(line)
        .and_then(|caps| caps.get(1))
        .map(|m| format!("http://{}", m.as_str()))
}

fn extract_localtunnel(line: &str) -> Option<String> {
    LOCALTUNNEL_RE
        .captures(line)
        .and_then(|caps| caps.get(1))
        .map(|m| m.as_str().to_string())
}

const PROVIDERS: &[Provider] = &[
    Provider {
        binary: "cloudflared",
        command: cloudflared_cmd,
        extract_url: extract_cloudflared,
    },
    Provider {
        binary: "ngrok",
        command: ngrok_cmd,
        extract_url: extract_ngrok,
    },
    Provider {
        binary: "bore",
        command: bore_cmd,
        extract_url: extract_bore,
    },
    Provider {
        binary: "lt",
        command: localtunnel_cmd,
        extract_url: extract_localtunnel,
    },
];

// ── Tunnel ──────────────────────────────────────────────────────────

/// Manage a public tunnel to a local port.
///
/// Implements `Drop` so the tunnel is always cleaned up.
pub struct Tunnel {
    process: Option<Child>,
    pub public_url: Option<String>,
}

impl Tunnel {
    pub fn new() -> Self {
        Tunnel {
            process: None,
            public_url: None,
        }
    }

    /// Start a tunnel and return the public URL.
    pub fn start(&mut self, local_port: u16) -> Result<String, TunnelError> {
        for provider in PROVIDERS {
            if which::which(provider.binary).is_err() {
                continue;
            }

            let bin_path = provider.binary;
            let cmd = (provider.command)(bin_path, local_port);

            match self.try_provider(cmd, provider.extract_url) {
                Ok(url) => {
                    self.public_url = Some(url.clone());
                    return Ok(url);
                }
                Err(_) => {
                    // Try next provider
                    self.stop();
                    continue;
                }
            }
        }

        Err(TunnelError {
            message: format!(
                "No tunnel tool found. Install one:\n\
                 \x20 macOS:  brew install cloudflared\n\
                 \x20 Linux:  curl -L https://github.com/cloudflare/cloudflared/releases/latest/download/cloudflared-linux-amd64 -o /usr/local/bin/cloudflared && chmod +x /usr/local/bin/cloudflared\n\
                 \x20 Or:     brew install ngrok && ngrok config add-authtoken <token>"
            ),
        })
    }

    fn try_provider(
        &mut self,
        cmd: Vec<String>,
        extract_url: fn(&str) -> Option<String>,
    ) -> Result<String, TunnelError> {
        let mut child = Command::new(&cmd[0])
            .args(&cmd[1..])
            .stdout(Stdio::piped())
            .stderr(Stdio::null())
            .spawn()
            .map_err(|e| TunnelError {
                message: format!("Failed to start tunnel: {}", e),
            })?;

        let stdout = child.stdout.take().ok_or_else(|| TunnelError {
            message: "Failed to capture stdout".to_string(),
        })?;

        let reader = BufReader::new(stdout);
        let deadline = Instant::now() + Duration::from_secs(15);

        let mut url: Option<String> = None;
        for line in reader.lines() {
            if Instant::now() > deadline {
                break;
            }
            match line {
                Ok(l) => {
                    if let Some(found) = extract_url(&l) {
                        url = Some(found);
                        break;
                    }
                }
                Err(_) => break,
            }

            // Check if process already died
            if child.try_wait().ok().flatten().is_some() {
                break;
            }
        }

        self.process = Some(child);

        url.ok_or_else(|| TunnelError {
            message: "Tunnel did not provide a public URL within 15 seconds".to_string(),
        })
    }

    /// Stop the tunnel process.
    pub fn stop(&mut self) {
        if let Some(ref mut child) = self.process {
            let _ = child.kill();
            let _ = child.wait();
        }
        self.process = None;
    }
}

impl Drop for Tunnel {
    fn drop(&mut self) {
        self.stop();
    }
}
